### ISSUE TITLE
Equivalent to click context

### ISSUE THREAD

**Issue by vikigenius** [BrianPugh/cyclopts#243]
Is there any equivalent to the click Context parameter and pass_context functionality?

https://click.palletsprojects.com/en/stable/complex/

**Comment by BrianPugh** [comment]
Not directly; what behavior are you trying to achieve? It can probably be accomplished by a different mechanism.

**Comment by vikigenius** [comment]
Mostly context and state management across commands and sub-commands instead of using globals/singletons. Eg: For DB management I can initialize the DB connection store it in the context object and pass it to sub commands for manipulations

**Comment by BrianPugh** [comment]
Does this [section of the meta app docs](https://cyclopts.readthedocs.io/en/latest/meta_app.html#custom-command-invocation) achieve what you are looking for?

**Comment by vikigenius** [comment]
Yeah the final section with the user example is what I want in a specific way. But reading that page it's not very clear what meta actually is.

Like for instance I am very clear on what the context object is and how it's passed between commands.

It seems like meta allows you to redefine how the cli and subcommands are launched.

And the example shows how you can leverage that to pass additional arguments.

But what happens when some of my commands need a user object and some don't.

This seems more cumbersome to pass through nested subcommands

**Comment by BrianPugh** [comment]
>  But reading that page it's not very clear what meta actually is.

Noted! I'll try and update the docs to make it more clear. But basically, the `meta` app is just another Cyclopt's application who's help page sort of gets merged in with the parenting cyclopts `App`.

Would it help if `parse_args` also returned a dictionary mapping of `parse=False` python-variable-names to their annotated type? That way you can:

```
@app.meta.default
def launcher(*tokens: Annotated[str, Parameter(show=False, allow_leading_hyphen=True)], user: str):
    additional_kwargs = {}
    command, bound, unparsed_variables = app.parse_args(tokens)
    # unparsed_variables would be {"user_obj": User}
    if "user_obj" in unparsed_variables:
        additional_kwargs["user_obj"] = User(user)
    return command(*bound.args, **bound.kwargs, **additional_kwargs)
```

I'm open to any proposed solution that makes the code terser/easer-to-read.

> This seems more cumbersome to pass through nested subcommands

In the above example, the `command` is fully resolved through nested subcommands, so handling them is no different.

**Comment by BrianPugh** [comment]
@vikigenius  Assuming the thumbsup is agreement for the `parse_args()` modification, I can implement this breaking change in the upcoming v3 release! I myself found the syntax without it a bit cumbersome (directly relying on `inspect.signature` to automate it), so I'm fine with this change. The only thing that makes me hesitant is if there is an alternative, better, cleaner way to do it ;).

**Comment by vikigenius** [comment]
Yeah I am thinking along similar lines, it does get me to what I want and it seems good to have this anyways, but the usage still seems a bit complex and hardcoded, maybe there is a better way, but I can't think of one now.

If I am reading the example correctly

`command, bound, unparsed_variables = app.parse_args(tokens)`

unparsed_variables would only contain `user_obj` if the corresponding command actually accepts that as an argument?

So for example if I had a delete command that doesn't take a user_obj

```
@app.command
def delete(
    user_id: int,
):
    print(f"Deleting User with ID: {user_id}")
```

The corresponding unparsed_variables would be empty?

But for this one:

```

@app.command
def create(
    age: int,
    *,
    user_obj: Annotated[User, Parameter(parse=False)],
):
    print(f"Creating user {user_obj.name} with age {age}.")
```

We would have the 'user_obj' key and it would map to the type User, why not just instantiate it using that

```
additional_kwargs["user_obj"] = unparsed_variables["user_obj"](user)
```

**Comment by BrianPugh** [comment]
> unparsed_variables would only contain user_obj if the corresponding command actually accepts that as an argument?

correct, more specifically it will contain any variable in the resulting parsed command's signature that has `Parameter(parse=False)`.

And the rest of your comment/example is exactly correct. A slight detail is that I think the mapped type would still be the fully Annotated type, so `user_obj: Annotated[User, Parameter(parse=False)]`. Even though it works, its technically incorrect to instantiate an object from an annotated type. A slight tweak on the idea is that I can have Cyclopts resolve the Annotation, so it would be `unparsed_variables == {"user_obj": User}`.

The overall solution is a bit clunky, but I'm afraid that any alternative solution I can think of is equally clunky, but would require the user to learn **yet another** thing about Cyclopts. I feel like with this approach the user still "feels in control" and there isn't too much unintuitive magic going on behind the scene or another method/function/feature to learn.

I personally am not a huge fan of the Click Context solution, because:
1. The special `obj` attribute is kind of C-like, where a library's `struct` may give you a `void *stick_whatever_data_you_want_here` for callbacks and similar. In C it's unavoidable, but it doesn't feel very nice in Python.
2. It requires you to understand click's `group` mechanism. Cyclopts doesn't really have a win here, as you have to understand (to some extent) Cyclopt's meta app system. However, within the meta app it's very clear what function is being executed and with what arguments.

**Comment by BrianPugh** [comment]
this is now implemented in `dev-3.0.0` as described above. Currently I have cyclopts resolving the annotation for the returned mapping.

**Comment by BrianPugh** [comment]
v3.0.0 has been released with this feature. Closing this issue for now. If you need anything else, please feel free to open another issue!
